compiler/rustc_typeck/src/collect.rs

```diff
@@ -984,54 +1029,156 @@ fn adt_def(tcx: TyCtxt<'_>, def_id: DefId) -> &ty::AdtDef {
+        if assoc_name.is_none() {
+            // FIXME: move this into the `super_predicates_of` query
+            for &(pred, span) in superbounds {
+                debug!("superbound: {:?}", pred);
+                if let ty::PredicateAtom::Trait(bound, _) = pred.skip_binders() {
+                    tcx.at(span).super_predicates_of(bound.def_id());
+                }
+            }
+        }
+
+        ty::GenericPredicates { parent: None, predicates: superbounds }
+    } else {
+        // if `assoc_name` is None, then the query should've been redirected to an
+        // external provider
+        assert!(assoc_name.is_some());
+        tcx.super_predicates_of(trait_def_id)
+    }
+}
+
+/// Ensures that the super traits of the trait with a `DefId`
+/// of `trait_def_id` are converted and stored. This also ensures that
+/// the transitive super traits are converted.
+fn super_traits_of(tcx: TyCtxt<'_>, trait_def_id: DefId) -> FxHashSet<DefId> {
+    let mut set = FxHashSet::default();
+    let mut stack = vec![trait_def_id];
+    while let Some(trait_did) = stack.pop() {
+        if !set.insert(trait_did) {
+            continue;
+        }
+
+        if trait_did.is_local() {
+            let trait_hir_id = tcx.hir().local_def_id_to_hir_id(trait_did.expect_local());
+
+            let item = match tcx.hir().get(trait_hir_id) {
+                Node::Item(item) => item,
+                _ => bug!("super_trait_of {} is not an item", trait_hir_id),
+            };
+
+            let (generics, supertraits) = match item.kind {
+                hir::ItemKind::Trait(.., ref generics, ref supertraits, _) => {
+                    (generics, supertraits)
+                }
+                hir::ItemKind::TraitAlias(ref generics, ref supertraits) => (generics, supertraits),
+                _ => span_bug!(item.span, "super_predicates invoked on non-trait"),
+            };
+
+            for supertrait in supertraits.iter() {
+                let trait_ref = supertrait.trait_ref();
+                if let Some(trait_did) = trait_ref.and_then(|trait_ref| trait_ref.trait_def_id()) {
+                    stack.push(trait_did);
+                }
+            }
+
+            let icx = ItemCtxt::new(tcx, trait_did);
+            // Convert any explicit superbounds in the where-clause,
+            // e.g., `trait Foo where Self: Bar`.
+            // In the case of trait aliases, however, we include all bounds in the where-clause,
+            // so e.g., `trait Foo = where u32: PartialEq<Self>` would include `u32: PartialEq<Self>`
+            // as one of its "superpredicates".
+            let is_trait_alias = tcx.is_trait_alias(trait_did);
+            let self_param_ty = tcx.types.self_param;
+            for (predicate, _) in icx.type_parameter_bounds_in_generics(
+                generics,
+                item.hir_id,
+                self_param_ty,
+                OnlySelfBounds(!is_trait_alias),
+                None,
+            ) {
```

Comment:
Ugh, such code duplication. I'm not sure the best way to fix it though. :(

Reply:
Yeah this is not great.